### PR TITLE
Convert suggestion response parsing to use NamedXContentRegistry

### PR DESCRIPTION
**Pull request by javanna**
We recently added parsing code to parse suggesters responses into java api objects. This was done using a switch based on the type of the returned suggestion. We can now replace the switch with using NamedXContentRegistry, which will also be used for aggs parsing.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/suggest/Suggest.java [hunk 1]
```diff
@@ -399,35 +400,17 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                         "Cannot parse suggestion response without type information. Set [" + RestSearchAction.TYPED_KEYS_PARAM
                                 + "] parameter on the request to ensure the type information is added to the response output");
             }
-            Suggestion suggestion = null;
-            Function<XContentParser, Entry> entryParser = null;
-            // the "size" parameter and the SortBy for TermSuggestion cannot be parsed from the response, use default values
-            // TODO investigate if we can use NamedXContentRegistry instead of this switch
-            switch (type) {
-                case Suggestion.NAME:
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: Should we keep support for this "generic" suggestion type as long as it can (in theory) still be instantiated? I'd be in favour of making it abstract and removing it from the transport layer (including support for its use in Custom Suggester) on master in a follow up PR after re-checking with @areek that we realy don't need it.

**Review by cbuescher — COMMENTED** [review]
Looks great, but I wonder if we should keep the parsing support for the top level Suggestion class and remove it/make it abstract in a follow up PR. I'm pretty confident we don't need it anymore but maybe we better recheck this and remove in a dedicated PR.

File: core/src/main/java/org/elasticsearch/search/suggest/Suggest.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: I do not understand what this is for. you can have a custom suggester but it has to be called "suggestion"? This makes no sense to me. If it is possible to register custom suggesters (I hear it may not be), RestHighLevelClient has to be subclassed and the custom parser has to be added to the registry, that's it. Note that these parsers will be added to the registry in a followup PR, as that change will go only to master.

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: With my limited knowledge of suggesters, I tend to think that we can remove the "generic" parsing case: we're not sure if custom suggesters really need this and even if it's the case a custom parser will have to be registered in the named xcontent registry anyway.

**Review by tlrx — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/search/suggest/Suggest.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: Okay, I was only suggesting that if we don't support it on the parsing side, we might as well remove it everywhere (it's use in CustomSuggester, making Suggestion class abstract, removing from transport layer etc...) and then it would probably be better to do this in a separate PR. I'm okay with not supporting the parser.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: yes we should totally clean that up. I was just against parsing back something that doesn't make sense :)

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/suggest/Suggest.java]: I was wrong when I said there are no custom Suggester implementations out there. I found this closed issue https://github.com/elastic/elasticsearch/issues/17100 where we discussed removing the the ability to register custom suggesters, which we didn't do because of plugin projects like this out there: https://github.com/meltwater/es-prefix-suggester. 
When I understand their code right, they only modify the Suggester implementation, which in this case outputs instances of the top level Suggestion type: /blob/master/src/main/java/com/meltwater/elasticsearch/search/suggest/PrefixSuggester.java
One could argue that they need to provide their own parser, but since they use our built-inSuggestion class it might also be reasonable if users assumed this gets properly parsed. Wdyt?
